app_helpers.py

```diff
@@ -110,3 +110,32 @@ def compress(response):
         return f(*args, **kwargs)
 
     return compressor
+
+
+def auth_admin_only(func):
+    """Test authentication on the request.
+    The request session should have previously authenticatated as an admin."""
+
+    @wraps(func)
+    def wrapper(*args, **kwargs):
+        if "username" in flask.session and flask.session["username"] is not None:
+            return func(*args, **kwargs)
+        else:
+            return flask.Response(response="Unauthorized", status=401)
```

Comment:
Have you tested this with the library registry admin? What happens when a session expires? Does the admin interface correctly handle getting a 401 and asking the user to log back in, or does the admin app just stop working?

Reply:
The admin interface is an SPA so it stays on the same page until refreshed, at which point the registration page shows up again.
The API requests to admin endpoints do fail in the meantime, which is something that should be addressed on the frontend.